### For all Normalized Training Tests
- Random, Varying, Increasing and Decreasing logic was changed.
- Epochs were reduced to 1000 max
- configs also changed

Find the best MLP of each type.
Find the best MLP overall.


In [1]:
import pandas as pd

In [9]:
# The filenames
varyMLP = "varyMLP_results01.txt"
staticMLP = "staticMLP_results01.txt"
increaseMLP = "increaseMLP_results01.txt"
decreaseMLP = "decreaseMLP_results01.txt"

# Random sizing, normalized data

In [10]:
the_output_text = open(varyMLP,'r')
read_output_text = the_output_text.read()
the_output_text.close()

In [11]:
print(len(read_output_text))
rezsplit_a = read_output_text.split("\n")
print(len(rezsplit_a))

989284
37


In [12]:
# Last line is just an empty new line. We remove it
print(rezsplit_a[36])
rezsplit_a.remove('')
print(len(rezsplit_a))


36


In [13]:
result_dict_list = [eval(rez) for rez in rezsplit_a]
print(len(result_dict_list))

36


In [14]:
print(result_dict_list[0].keys())
print(result_dict_list[0]["config"])
print(min(result_dict_list[0]["result"]["loss"]))
print(result_dict_list[0]["result"]["loss"][-1])

dict_keys(['epochs', 'config', 'result'])
{'layers': 1, 'layer': {1: 32}}
0.0001868045348643322
0.00020166372093863833


Since overtraiing can be an issue, it's better to get the last MSE loss value, instead of the minimum MSE loss value, per network trained. This last value represents the minimized loss after all epochs, and thus, is subject to overfitting. The winner of all the trained networks would have exhibited the least overfitting (forces MSE up) but also have the lowest score.

In [15]:
MSE_loss_results = [{"epochs": item['epochs'],
                    "config":item["config"],
                    "last_loss": item["result"]["loss"][-1]} for item in result_dict_list]
print(MSE_loss_results[0])

{'epochs': 250, 'config': {'layers': 1, 'layer': {1: 32}}, 'last_loss': 0.00020166372093863833}


In [16]:
ALL_last_loss = [d["last_loss"] for d in MSE_loss_results]
print(len(ALL_last_loss))
print(min(ALL_last_loss))
ALL_last_loss[:5]

36
0.0001354424024335794


[0.00020166372093863833,
 0.00022402345081837388,
 0.00020199551104134556,
 0.00021293265378124175,
 0.0001909192095367736]

In [17]:
sorted_losses = ALL_last_loss.copy()
sorted_losses.sort()
sorted_losses[:10]

[0.0001354424024335794,
 0.00014373036288282075,
 0.00014397148111834737,
 0.00014799091524656603,
 0.00015185584267390055,
 0.0001522503757123949,
 0.00015372951181141334,
 0.0001563710935118765,
 0.00015762507575087668,
 0.00016117272400043646]

Above, we have the 10 lowest MSE losses amongst the __36__ trained networks.  
What we will do now, if go back and find the epochs used to train these networks, and their associated architecture.

In [18]:
top_ten_losses = sorted_losses[:10]
top_ten_networks = [ntwk for ntwk in MSE_loss_results if ntwk["last_loss"] in top_ten_losses]
top_ten_networks

[{'epochs': 500,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.0001563710935118765},
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00015185584267390055},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 90}},
  'last_loss': 0.00014397148111834737},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 32, 2: 40}},
  'last_loss': 0.00015372951181141334},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 80}},
  'last_loss': 0.00016117272400043646},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.0001522503757123949},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014373036288282075},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 60}},
  'last_loss': 0.00014799091524656603},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_l

### (Old) WInner - Randomized size, Normalized Data
`{'epochs': 5000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 80, 3: 64}},
  'last_loss': 0.0001037829471429312}`

### WInner - Randomized size, Normalized Data
`{'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 80, 3: 60, 4: 45}},
  'last_loss': 0.0001354424024335794}`
  
runner up  
`{'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014373036288282075}`

third place  
`{'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 90}},
  'last_loss': 0.00014397148111834737}`

In [21]:
#old
0.00012337757427135495/0.0001037829471429312

1.1888039188310753

In [20]:
0.00014373036288282075/0.0001354424024335794

1.0611917708215914

In [22]:
0.00014397148111834737/0.0001354424024335794

1.0629719979232546

The winning network used 1000 epochs. 
- The best 500 epoch network had an MSE 6.297% higher.
- The next best network overall had an MSE 6.119% higher (1000 epochs)

# Decreasing Size, Normalized Data

In [23]:
dec_output_text = open(decreaseMLP,'r')
read_output_text = dec_output_text.read()
dec_output_text.close()

In [24]:
print(len(read_output_text))
rezsplit_b = read_output_text.split("\n")
print(len(rezsplit_b))

989295
37


In [25]:
# Last time is just an empty new line. We remove it
print(rezsplit_b[36].__repr__())
rezsplit_b.remove('')
print(len(rezsplit_b))

''
36


In [26]:
dec_result_dict_list = [eval(rez) for rez in rezsplit_b]
print(len(dec_result_dict_list))

36


In [27]:
print(dec_result_dict_list[0].keys())
print(dec_result_dict_list[0]["config"])
print(min(dec_result_dict_list[0]["result"]["loss"]))
print(dec_result_dict_list[0]["result"]["loss"][-1])

dict_keys(['epochs', 'config', 'result'])
{'layers': 1, 'layer': {1: 32}}
0.0001889552990502521
0.00019599481484055113


Since overtraiing can be an issue, it's better to get the last MSE loss value, instead of the minimum MSE loss value, per networl trained. This last value represents the minimized loss after all epochs, and thus, is subject to overfitting. The winner of all the trained networks would have exhibited the least overfitting (forces MSE up) but also have the lowest score.

In [28]:
dec_MSE_loss_results = [{"epochs": item['epochs'],
                    "config":item["config"],
                    "last_loss": item["result"]["loss"][-1]} for item in dec_result_dict_list]
print(dec_MSE_loss_results[0])

{'epochs': 250, 'config': {'layers': 1, 'layer': {1: 32}}, 'last_loss': 0.00019599481484055113}


In [29]:
ALL_dec_last_loss = [d["last_loss"] for d in dec_MSE_loss_results]
print(len(ALL_dec_last_loss))
print(min(ALL_dec_last_loss))
ALL_dec_last_loss[:5]

36
0.00014170785411491546


[0.00019599481484055113,
 0.00019372542065617448,
 0.00020687423349456498,
 0.00021023944865723168,
 0.00017990735303779616]

In [30]:
sorted_declosses = ALL_dec_last_loss.copy()
sorted_declosses.sort()
sorted_declosses[:10]

[0.00014170785411491546,
 0.0001446379672043585,
 0.0001453293824861158,
 0.0001500321141726156,
 0.00015136461343140596,
 0.00015831690950825558,
 0.00015963765814352516,
 0.00016071098382994318,
 0.00016386865501514123,
 0.00016402309331481989]

Above, we have the 10 lowest MSE losses amongst the __36__ trained networks.  
What we will do now, if go back and find the epochs used to train these networks, and their associated architecture.

In [32]:
dec_top_ten_losses = sorted_declosses[:10]
dec_top_ten_networks = [ntwk for ntwk in dec_MSE_loss_results if ntwk["last_loss"] in dec_top_ten_losses]
dec_top_ten_networks

[{'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00016386865501514123},
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_loss': 0.0001453293824861158},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 32, 2: 24, 3: 18, 4: 13}},
  'last_loss': 0.00015831690950825558},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 32}},
  'last_loss': 0.00016402309331481989},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 64}},
  'last_loss': 0.00015963765814352516},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.00015136461343140596},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00014170785411491546},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 32, 2: 24, 3: 18, 4: 13}},
  'last_loss': 0.00016071098382994318},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 48, 3: 36, 4: 27}},
  '

### (OLD) Winner, decreasing only, normalized data
`{'epochs': 2500,
  'config': {'layers': 4, 'layer': {1: 64, 2: 48, 3: 48, 4: 48}}`

### Winner, decreasing only, normalized data
`{'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00014170785411491546}`
  
runner up  
`{'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 54}},
  'last_loss': 0.0001446379672043585}`

third place  
`{'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_loss': 0.0001453293824861158}`

# Increasing Size, Normalized Data

In [34]:
inc_output_text = open(increaseMLP,'r')
read_output_text = inc_output_text.read()
inc_output_text.close()

In [35]:
print(len(read_output_text))
rezsplit_b = read_output_text.split("\n")
print(len(rezsplit_b))

989551
37


In [36]:
# Last time is just an empty new line. We remove it
print(rezsplit_b[36].__repr__())
rezsplit_b.remove('')
print(len(rezsplit_b))

''
36


In [37]:
inc_result_dict_list = [eval(rez) for rez in rezsplit_b]
print(len(inc_result_dict_list))

36


In [40]:
print(inc_result_dict_list[0].keys())
print(inc_result_dict_list[0]["config"])
print(min(inc_result_dict_list[0]["result"]["loss"]))
print(inc_result_dict_list[0]["result"]["loss"][-1])

dict_keys(['epochs', 'config', 'result'])
{'layers': 1, 'layer': {1: 32}}
0.0001851988639071639
0.00020339362779148108


Since overtraiing can be an issue, it's better to get the last MSE loss value, instead of the minimum MSE loss value, per networl trained. This last value represents the minimized loss after all epochs, and thus, is subject to overfitting. The winner of all the trained networks would have exhibited the least overfitting (forces MSE up) but also have the lowest score.

In [41]:
inc_MSE_loss_results = [{"epochs": item['epochs'],
                    "config":item["config"],
                    "last_loss": item["result"]["loss"][-1]} for item in inc_result_dict_list]
print(inc_MSE_loss_results[0])

{'epochs': 250, 'config': {'layers': 1, 'layer': {1: 32}}, 'last_loss': 0.00020339362779148108}


In [42]:
ALL_inc_last_loss = [d["last_loss"] for d in inc_MSE_loss_results]
print(len(ALL_inc_last_loss))
print(min(ALL_inc_last_loss))
ALL_inc_last_loss[:5]

36
0.00013302507601236916


[0.00020339362779148108,
 0.0001915324802653366,
 0.00018771469359929295,
 0.0002014458056719107,
 0.0001930927582278774]

In [43]:
sorted_inclosses = ALL_inc_last_loss.copy()
sorted_inclosses.sort()
sorted_inclosses[:10]

[0.00013302507601236916,
 0.0001337122449564145,
 0.00013957328457775658,
 0.00014199366323917293,
 0.00014273140720850956,
 0.00014613495675040863,
 0.00014732041913538053,
 0.000151243927698773,
 0.0001517431779761519,
 0.00015306539593463097]

Above, we have the 10 lowest MSE losses amongst the __36__ trained networks.  
What we will do now, if go back and find the epochs used to train these networks, and their associated architecture.

In [44]:
inc_top_ten_losses = sorted_inclosses[:10]
inc_top_ten_networks = [ntwk for ntwk in inc_MSE_loss_results if ntwk["last_loss"] in inc_top_ten_losses]
inc_top_ten_networks

[{'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 32, 2: 40, 3: 50, 4: 62}},
  'last_loss': 0.000151243927698773},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 64, 2: 80, 3: 100, 4: 125}},
  'last_loss': 0.0001517431779761519},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 80}},
  'last_loss': 0.00013957328457775658},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 160}},
  'last_loss': 0.00014732041913538053},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014613495675040863},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 80, 3: 100}},
  'last_loss': 0.0001337122449564145},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 160, 3: 200}},
  'last_loss': 0.00015306539593463097},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 32, 2: 40, 3: 50, 4: 62}},
  'last_loss': 0.00014273140720850956},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1

# Static Size, Normalized Data

In [47]:
fix_output_text = open(staticMLP,'r')
read_output_text = fix_output_text.read()
fix_output_text.close()

In [48]:
print(len(read_output_text))
rezsplit_b = read_output_text.split("\n")
print(len(rezsplit_b))

989219
37


In [49]:
# Last time is just an empty new line. We remove it
print(rezsplit_b[36].__repr__())
rezsplit_b.remove('')
print(len(rezsplit_b))

''
36


In [50]:
fix_result_dict_list = [eval(rez) for rez in rezsplit_b]
print(len(fix_result_dict_list))

36


In [51]:
print(fix_result_dict_list[0].keys())
print(fix_result_dict_list[0]["config"])
print(min(fix_result_dict_list[0]["result"]["loss"]))
print(fix_result_dict_list[0]["result"]["loss"][-1])

dict_keys(['epochs', 'config', 'result'])
{'layers': 1, 'layer': {1: 32}}
0.00018686378819801885
0.00020834982382493222


Since overtraiing can be an issue, it's better to get the last MSE loss value, instead of the minimum MSE loss value, per networl trained. This last value represents the minimized loss after all epochs, and thus, is subject to overfitting. The winner of all the trained networks would have exhibited the least overfitting (forces MSE up) but also have the lowest score.

In [52]:
fix_MSE_loss_results = [{"epochs": item['epochs'],
                    "config":item["config"],
                    "last_loss": item["result"]["loss"][-1]} for item in fix_result_dict_list]
print(fix_MSE_loss_results[0])

{'epochs': 250, 'config': {'layers': 1, 'layer': {1: 32}}, 'last_loss': 0.00020834982382493222}


In [53]:
ALL_fix_last_loss = [d["last_loss"] for d in fix_MSE_loss_results]
print(len(ALL_fix_last_loss))
print(min(ALL_fix_last_loss))
ALL_fix_last_loss[:5]

36
0.00013208114798526547


[0.00020834982382493222,
 0.00020161756015038456,
 0.00019288259567408562,
 0.00019946300622310933,
 0.00018860418520477468]

In [54]:
sorted_fixlosses = ALL_fix_last_loss.copy()
sorted_fixlosses.sort()
sorted_fixlosses[:10]

[0.00013208114798526547,
 0.0001338472487890967,
 0.00013810502170101375,
 0.0001401019461488308,
 0.00014958753438417495,
 0.0001532250480582667,
 0.00015711002483104875,
 0.00015932932335087298,
 0.00016036911959961216,
 0.000160697456095798]

Above, we have the 10 lowest MSE losses amongst the __36__ trained networks.  
What we will do now, if go back and find the epochs used to train these networks, and their associated architecture.

In [55]:
fix_top_ten_losses = sorted_fixlosses[:10]
fix_top_ten_networks = [ntwk for ntwk in fix_MSE_loss_results if ntwk["last_loss"] in fix_top_ten_losses]
fix_top_ten_networks

[{'epochs': 250,
  'config': {'layers': 3, 'layer': {1: 64, 2: 64, 3: 64}},
  'last_loss': 0.00015932932335087298},
 {'epochs': 500,
  'config': {'layers': 2, 'layer': {1: 128, 2: 128}},
  'last_loss': 0.0001532250480582667},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 64, 2: 64, 3: 64, 4: 64}},
  'last_loss': 0.000160697456095798},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 128}},
  'last_loss': 0.00016036911959961216},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 32, 2: 32}},
  'last_loss': 0.00015711002483104875},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 64}},
  'last_loss': 0.00014958753438417495},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 32, 3: 32}},
  'last_loss': 0.0001401019461488308},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 64, 3: 64}},
  'last_loss': 0.00013208114798526547},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 128, 3: 128}},
  'last_loss': 0

# ALL

In [59]:
top_ten_networks

[{'epochs': 500,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.0001563710935118765},
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00015185584267390055},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 90}},
  'last_loss': 0.00014397148111834737},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 32, 2: 40}},
  'last_loss': 0.00015372951181141334},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 80}},
  'last_loss': 0.00016117272400043646},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.0001522503757123949},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014373036288282075},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 60}},
  'last_loss': 0.00014799091524656603},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_l

In [60]:
inc_top_ten_networks

[{'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 32, 2: 40, 3: 50, 4: 62}},
  'last_loss': 0.000151243927698773},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 64, 2: 80, 3: 100, 4: 125}},
  'last_loss': 0.0001517431779761519},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 80}},
  'last_loss': 0.00013957328457775658},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 160}},
  'last_loss': 0.00014732041913538053},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014613495675040863},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 80, 3: 100}},
  'last_loss': 0.0001337122449564145},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 160, 3: 200}},
  'last_loss': 0.00015306539593463097},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 32, 2: 40, 3: 50, 4: 62}},
  'last_loss': 0.00014273140720850956},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1

In [61]:
dec_top_ten_networks

[{'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00016386865501514123},
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_loss': 0.0001453293824861158},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 32, 2: 24, 3: 18, 4: 13}},
  'last_loss': 0.00015831690950825558},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 32}},
  'last_loss': 0.00016402309331481989},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 64}},
  'last_loss': 0.00015963765814352516},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.00015136461343140596},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00014170785411491546},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 32, 2: 24, 3: 18, 4: 13}},
  'last_loss': 0.00016071098382994318},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 48, 3: 36, 4: 27}},
  '

In [62]:
fix_top_ten_networks

[{'epochs': 250,
  'config': {'layers': 3, 'layer': {1: 64, 2: 64, 3: 64}},
  'last_loss': 0.00015932932335087298},
 {'epochs': 500,
  'config': {'layers': 2, 'layer': {1: 128, 2: 128}},
  'last_loss': 0.0001532250480582667},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 64, 2: 64, 3: 64, 4: 64}},
  'last_loss': 0.000160697456095798},
 {'epochs': 1000,
  'config': {'layers': 1, 'layer': {1: 128}},
  'last_loss': 0.00016036911959961216},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 32, 2: 32}},
  'last_loss': 0.00015711002483104875},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 64}},
  'last_loss': 0.00014958753438417495},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 32, 3: 32}},
  'last_loss': 0.0001401019461488308},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 64, 3: 64}},
  'last_loss': 0.00013208114798526547},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 128, 3: 128}},
  'last_loss': 0

Top 5 per search type
- Varying
`
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00015185584267390055},
 {'epochs': 500,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 90}},
  'last_loss': 0.00014397148111834737},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 40, 3: 50}},
  'last_loss': 0.00014373036288282075},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 60}},
  'last_loss': 0.00014799091524656603},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 80, 3: 60, 4: 45}},
  'last_loss': 0.0001354424024335794}`
- Increasing
`
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 80}},
  'last_loss': 0.00013957328457775658},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 80, 3: 100}},
  'last_loss': 0.0001337122449564145},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 32, 2: 40, 3: 50, 4: 62}},
  'last_loss': 0.00014273140720850956},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 80, 3: 100, 4: 125}},
  'last_loss': 0.00014199366323917293},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 128, 2: 160, 3: 200, 4: 250}},
  'last_loss': 0.00013302507601236916}`
- Decreasing
`
 {'epochs': 500,
  'config': {'layers': 3, 'layer': {1: 128, 2: 96, 3: 72}},
  'last_loss': 0.0001453293824861158},
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 128, 2: 96}},
  'last_loss': 0.00015136461343140596},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 48, 3: 36}},
  'last_loss': 0.00014170785411491546},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 64, 2: 48, 3: 36, 4: 27}},
  'last_loss': 0.0001500321141726156},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 128, 2: 96, 3: 72, 4: 54}},
  'last_loss': 0.0001446379672043585}`
- Static
`
 {'epochs': 1000,
  'config': {'layers': 2, 'layer': {1: 64, 2: 64}},
  'last_loss': 0.00014958753438417495},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 32, 2: 32, 3: 32}},
  'last_loss': 0.0001401019461488308},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 64, 2: 64, 3: 64}},
  'last_loss': 0.00013208114798526547},
 {'epochs': 1000,
  'config': {'layers': 3, 'layer': {1: 128, 2: 128, 3: 128}},
  'last_loss': 0.00013810502170101375},
 {'epochs': 1000,
  'config': {'layers': 4, 'layer': {1: 128, 2: 128, 3: 128, 4: 128}},
  'last_loss': 0.0001338472487890967}`